mypy/meet.py

```diff
@@ -51,155 +56,246 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
+    elif isinstance(right, TypedDictType):
+        right = right.fallback
+
+    if is_tuple(left) and is_tuple(right):
+        return are_tuples_overlapping(left, right, ignore_promotions=ignore_promotions)
+    elif isinstance(left, TupleType):
+        left = left.fallback
+    elif isinstance(right, TupleType):
+        right = right.fallback
+
+    # Next, we handle single-variant types that cannot be inherently partially overlapping,
+    # but do require custom logic to inspect.
+    #
+    # As before, we degrade into 'Instance' whenever possible.
+
+    if isinstance(left, TypeType) and isinstance(right, TypeType):
+        # TODO: Can Callable[[...], T] and Type[T] be partially overlapping?
+        return _is_overlapping_types(left.item, right.item)
+
+    if isinstance(left, CallableType) and isinstance(right, CallableType):
+        return is_callable_compatible(left, right,
+                                      is_compat=_is_overlapping_types,
+                                      ignore_pos_arg_names=True,
+                                      allow_partial_overlap=True)
+    elif isinstance(left, CallableType):
+        left = left.fallback
+    elif isinstance(right, CallableType):
+        right = right.fallback
+
+    # Finally, we handle the case where left and right are instances.
+
+    if isinstance(left, Instance) and isinstance(right, Instance):
+        if left.type.is_protocol and is_protocol_implementation(right, left):
+            return True
+        if right.type.is_protocol and is_protocol_implementation(left, right):
+            return True
+
+        # Two unrelated types cannot be partially overlapping: they're disjoint.
+        # We don't need to handle promotions because they've already been handled
+        # by the calls to `is_subtype(...)` up above (and promotable types never
+        # have any generic arguments we need to recurse on).
+        if left.type.has_base(right.type.fullname()):
+            left = map_instance_to_supertype(left, right.type)
+        elif right.type.has_base(left.type.fullname()):
+            right = map_instance_to_supertype(right, left.type)
         else:
-            return isinstance(other, CallableType) and is_subtype(other, one)
-    if experiments.STRICT_OPTIONAL:
-        if isinstance(t, NoneTyp) != isinstance(s, NoneTyp):
-            # NoneTyp does not overlap with other non-Union types under strict Optional checking
             return False
-    # We conservatively assume that non-instance, non-union, non-TupleType and non-TypeType types
-    # can overlap any other types.
+
+        if len(left.args) == len(right.args):
+            for left_arg, right_arg in zip(left.args, right.args):
+                if _is_overlapping_types(left_arg, right_arg):
+                    return True
+
+    # We ought to have handled every case by now: we conclude the
+    # two types are not overlapping, either completely or partially.
+
+    return False
```

Comment:
I think it would be safer to have different defaults for overloads and `isinstance()` checks. For the former, it is better to have `False` to avoid false positives. But for `isinstance()` it is probably better to have `True` for the cases where it is not clear (to avoid inferring some branches as unreachable, as this may cause problems for mypyc).

Reply:
Could I persuade you to keep this a `return False`? Basically the way `is_overlapping_types` works is by steadily working through each pair of types and comparing them against each other.

So, if the left and the right are both TypedDict, we compare those two directly and return either True or False, if the left and the right are both Instance, we do the same...

This final `return False` case is meant to handle the situation where left and right are genuinely two unrelated types -- returning anything other then False here would lead to incorrect results in those cases.

It's possible that there are some overlap cases we've missed (e.g. can CallableType overlap with Type? Can Type overlap with Instance?), but I'd much rather prefer we handle those missing cases directly and fix `is_overlapping_types` as they come up instead of silently letting the function continue to be imprecise.

> to avoid inferring some branches as unreachable, as this may cause problems for mypyc

I think maybe a better way of solving this is to make mypy report branches that are accidentally unreachable -- it's fairly easy to accidentally add an isinstance check that'll never be reached, I think. (Of course, we wouldn't report branches that are *intentionally* unreachable due to system platform guards or things like 'if False' -- we also probably shouldn't report branches that just contain an assert/just throw an exception).

I'm hoping that such a check could just be enabled by default -- I can't really imagine people intentionally adding logic to a no-op branch.

(I'd be willing to work on adding this sort of check if you think it's valuable enough).

Reply:
> Could I persuade you to keep this a `return False`?

Not just me :-)

Your proposal with the additional error may generate false positives. Consider this example:
```python
class A: ...
class B: ...
if isinstance(x, A):
    ...
    if isinstance(x, B):
        # this is not reachable ...or is it?
```
I think we already discussed some time ago that there are special cases (like multiple inheritance) where we are deliberately unsafe. All such cases should be treated on per-use-case basis. For example, when checking for overloads, we treat `A` and `B` as non-overlapping, while when inferring reachability of a branch we might want to do the opposite.

Even if you will not add the `default` argument for such corner cases, I am quite sure it will be added later (the above example is unreachable on master, but this may be changed soon).

@msullivan what do you think about this?

Reply:
> maybe a better approach might be to add a `allow_multiple_inheritance` flag to this function?
[snip]
> More broadly, I'd prefer whitelisting specific cases where we want to relax the overlap checks -- I think the code would be more maintainable that way

You think it is more maintainable, I think it looks like "premature maintainability optimization" :-) 

Currently all edge cases should be considered non-overlapping for overloads, but overlapping for inferring reachability. For example, empty containers: `List[int]` and `List[Type[int]]` are considered non-overlapping for overloads but should be overlapping for reachability.

Anyway, it looks like this is an independent issue, so could you please add a TODO item above the final `return False`, and add a permalink to it in https://github.com/python/mypy/issues/3603 ?

Reply:
> Currently all edge cases should be considered non-overlapping for overloads, but overlapping for inferring reachability. For example, empty containers: List[int] and List[Type[int]] are considered non-overlapping for overloads but should be overlapping for reachability.

Well, `List[int]` and `List[Type[int]]` are both represented by Instance objects underneath -- so I think special-casing the "left and right are both Instances" tactic would work?

In any case, I'll make a TODO -- we can worry about this later once it comes up again :)

Reply:
> so I think special-casing the "left and right are both Instances" tactic would work?

It may be more subtle because it will recurse to checking arguments that are `Instance` vs `TypeType`. Also there are other cases if you want non-instances, e.g. `Instance` vs `CallableType` should be _always_ overlapping for the purpose of `isinstance()` and `callable()` checks (that @msullivan implemented a while ago), but should not be overlapping for overloads.

> In any case, I'll make a TODO -- we can worry about this later once it comes up again :)

OK.